api/v1/storage_autoscaler.go

```diff
@@ -0,0 +1,142 @@
+	// Should be less than the OsdNearFullThresholdPercentage.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=70
+	StorageScalingThresholdPercent int `json:"storageScalingThresholdPercent,omitempty"`
+	// TimeoutSeconds is the time in seconds after which the storage auto-scaler will alert the user that the scaling operation has been failed.
+	// +kubebuilder:validation:Optional
+	// +kubebuilder:default=1800
+	TimeoutSeconds int `json:"timeoutSeconds,omitempty"`
+}
+
+type StorageAutoScalerPhase string
+
+const (
+	StorageAutoScalerPhaseNotStarted StorageAutoScalerPhase = "NotStarted"
+	StorageAutoScalerPhaseInProgress StorageAutoScalerPhase = "InProgress"
+	StorageAutoScalerPhaseSucceeded  StorageAutoScalerPhase = "Succeeded"
+	StorageAutoScalerPhaseFailed     StorageAutoScalerPhase = "Failed"
+)
+
+// TimestampedError is the error message with the time stamp.
+type TimestampedError struct {
+	// Message is the error message in case the storage scaling operation has failed.
+	Message string `json:"message,omitempty"`
+	// Timestamp is the time stamp when the error occurred.
+	Timestamp metav1.Time `json:"timestamp,omitempty"`
+}
+
+type LastExpansionStatus struct {
+	// StartTime is the time stamp of the last run start of the storage scaling
+	// +optional
+	StartTime metav1.Time `json:"startTime,omitempty"`
+	// CompletionTime is the time stamp of the last run completion of the storage scaling
+	// +optional
+	CompletionTime metav1.Time `json:"completionTime,omitempty"`
+	// The start OSD size is the original size of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected OSD size.
+	// Used for vertical scaling of OSDs.
+	// +optional
+	StartOsdSize resource.Quantity `json:"startOsdSize,omitempty"`
+	// The ExpectedOsdSize is the size that the auto-expansion has decided to set.
+	// This will be set on the storageCLuster CR as the desired size of the OSDs.
+	// Used for vertical scaling of OSDs.
+	// +optional
+	ExpectedOsdSize resource.Quantity `json:"expectedOsdSize,omitempty"`
+	// The start OSD count is the original count of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected OSD count.
+	// Used for horizontal scaling of OSDs.
+	// +optional
+	StartOsdCount uint16 `json:"startOsdCount,omitempty"`
+	// The Expected OSD count is the count that the auto-expansion has decided to set.
+	// This will be set on the storageCluster CR as the desired count of the OSDs.
+	// Used for horizontal scaling of OSDs.
+	// +optional
+	ExpectedOsdCount uint16 `json:"expectedOsdCount,omitempty"`
+	// StorageCapacityLimitReached is the flag that indicates if the storage capacity limit has been reached.
+	StorageCapacityLimitReached bool `json:"storageCapacityLimitReached,omitempty"`
```

Comment:
Seems like this bool should go on the `StorageAutoScalerStatus` struct, rather than the last expansion status struct